src/coreclr/jit/helperexpansion.cpp

```diff
@@ -1553,3 +1602,317 @@ bool Compiler::fgVNBasedIntrinsicExpansionForCall_ReadUtf8(BasicBlock** pBlock,
+    //     if (tmp->pMT == likelyClass)
+    //         goto typeCheckSucceedBb;
+    //
+    // fallbackBb (BBJ_ALWAYS):                     [weight: <profile>]
+    //     tmp = helper_call(cls, tmp);
+    //     goto lastBlock;
+    //
+    // typeCheckSucceedBb (BBJ_ALWAYS):             [weight: <profile>]
+    //     nop (or tmp = null)
+    //
+    // lastBlock (BBJ_any):                         [weight: 1.0]
+    //     use(tmp);
+    //
+
+    // Block 1: nullcheckBb
+    // TODO-InlineCast: assertionprop should leave us a mark that objArg is never null, so we can omit this check
+    // it's too late to rely on upstream phases to do this for us (unless we do optRepeat).
+    GenTree* objTmp      = gtNewLclVarNode(tmpNum);
+    GenTree* nullcheckOp = gtNewOperNode(GT_EQ, TYP_INT, objTmp, gtNewNull());
+    nullcheckOp->gtFlags |= GTF_RELOP_JMP_USED;
+    BasicBlock* nullcheckBb = fgNewBBFromTreeAfter(BBJ_COND, firstBb, gtNewOperNode(GT_JTRUE, TYP_VOID, nullcheckOp),
+                                                   debugInfo, lastBb, true);
+
+    // Insert statements to spill call's arguments (preserving the evaluation order)
+    // TODO-InlineCast: don't spill cls if possible, we only need it after the nullcheck
+    const unsigned clsTmp   = lvaGrabTemp(true DEBUGARG("local for cls"));
+    lvaTable[clsTmp].lvType = clsArg->TypeGet();
+    Statement* storeObjStmt = fgNewStmtAtBeg(nullcheckBb, gtNewTempStore(tmpNum, objArg), debugInfo);
+    Statement* storeClsStmt = fgNewStmtAtBeg(nullcheckBb, gtNewTempStore(clsTmp, clsArg), debugInfo);
+    gtSetStmtInfo(storeObjStmt);
+    gtSetStmtInfo(storeClsStmt);
+    fgSetStmtSeq(storeObjStmt);
+    fgSetStmtSeq(storeClsStmt);
+    gtUpdateStmtSideEffects(storeObjStmt);
+    gtUpdateStmtSideEffects(storeClsStmt);
+
+    // if likelyCls == clsArg, we can just use clsArg that we've just spilled to a temp
+    // it's a sort of manual CSE.
+    GenTree* likelyClsNode = likelyCls == expectedCls ? gtNewLclVarNode(clsTmp) : gtNewIconEmbClsHndNode(likelyCls);
+
+    // Block 2: typeCheckBb
+    GenTree* mtCheck = gtNewOperNode(GT_EQ, TYP_INT, gtNewMethodTableLookup(gtCloneExpr(objTmp)), likelyClsNode);
+    mtCheck->gtFlags |= GTF_RELOP_JMP_USED;
+    GenTree*    jtrue       = gtNewOperNode(GT_JTRUE, TYP_VOID, mtCheck);
+    BasicBlock* typeCheckBb = fgNewBBFromTreeAfter(BBJ_COND, nullcheckBb, jtrue, debugInfo, lastBb, true);
+
+    // Block 3: fallbackBb
+    // NOTE: we spilled call's arguments above, we need to re-create it here (we can't just modify call's args)
+    GenTree* fallbackCall =
+        gtNewHelperCallNode(call->GetHelperNum(), TYP_REF, gtNewLclVarNode(clsTmp), gtNewLclVarNode(tmpNum));
+    fallbackCall = fgMorphCall(fallbackCall->AsCall());
+    gtSetEvalOrder(fallbackCall);
+    GenTree*    fallbackTree = gtNewTempStore(tmpNum, fallbackCall);
+    BasicBlock* fallbackBb   = fgNewBBFromTreeAfter(BBJ_ALWAYS, typeCheckBb, fallbackTree, debugInfo, lastBb, true);
```

Comment:
I'm not sure I understand why this needs to recreate the call. It probably doesn't after you remove the spill of the call arguments above?

Reply:
I'll remove the spill for now to simplify the code. But there is a possible situation where likelyclass matches casthelper's arg - in that case we want to CSE it by hands and replace the arg. Let's simplify for now and don't do it.

Reply:
Ok. Note that it's fine to change the args of a call node by modifying the late node if it has one, and otherwise the early node.

Reply:
Addressed